src/seq/coin_flipper.rs

```diff
@@ -0,0 +1,152 @@
+                // If n * 2^c > `usize::MAX` we always return `true` anyway
+                n = n.saturating_mul(2_usize.pow(c));
+            } else {
+                //At least one tail
+                if c == 1 {
+                    // Calculate 2n - d.
+                    // We need to use wrapping as 2n might be greater than `usize::MAX`
+                    let next_n = n.wrapping_add(n).wrapping_sub(d);
+                    if next_n == 0 || next_n > n {
+                        // This will happen if 2n < d
+                        return false;
+                    }
+                    n = next_n;
+                } else {
+                    // c > 1 so 2n < d so we can return false
+                    return false;
+                }
+            }
+        }
+        true
+    }
+
+    /// If the next `c` bits of randomness all represent heads, consume them, return true
+    /// Otherwise return false and consume the number of heads plus one.
+    /// Generates new bits of randomness when necessary (in 32 bit chunks)
+    /// Has a 1 in 2 to the `c` chance of returning true
+    /// `c` must be less than or equal to 32
+    fn flip_c_heads(&mut self, mut c: u32) -> bool {
+        debug_assert!(c <= 32);
+        // Note that zeros on the left of the chunk represent heads.
+        // It needs to be this way round because zeros are filled in when left shifting
+        loop {
+            let zeros = self.chunk.leading_zeros();
+
+            if zeros < c {
+                // The happy path - we found a 1 and can return false
+                // Note that because a 1 bit was detected,
+                // We cannot have run out of random bits so we don't need to check
+
+                // First consume all of the bits read
+                // Using shl seems to give worse performance for size-hinted iterators
+                self.chunk = self.chunk.wrapping_shl(zeros + 1);
+
+                self.chunk_remaining = self.chunk_remaining.saturating_sub(zeros + 1);
+                return false;
+            } else {
+                // The number of zeros is larger than `c`
+                // There are two possibilities
+                if let Some(new_remaining) = self.chunk_remaining.checked_sub(c) {
+                    // Those zeroes were all part of our random chunk,
+                    // throw away `c` bits of randomness and return true
+                    self.chunk_remaining = new_remaining;
+                    self.chunk <<= c;
+                    return true;
+                } else {
+                    // Some of those zeroes were part of the random chunk
+                    // and some were part of the space behind it
+                    // We need to take into account only the zeroes that were random
+                    c -= self.chunk_remaining;
+
+                    // Generate a new chunk
+                    self.chunk = self.rng.next_u32();
+                    self.chunk_remaining = 32;
```

Comment:
Using `next_u32` is sub-optimal when using a 64-bit native generator. We can't fix this now (except using `next_u64` instead, possibly worse depending on the RNG and CPU). We should probably leave a note like this:

> TODO(opt): this should depend on RNG word size

Relates to #1261: wanting a method to determine native output size.

Anyway, this shouldn't be a blocker, though unfortunately it means value-breaking changes again in the future are likely.

Reply:
Agreed. Though the performance difference isn't enormous, even when using a 64 bit rng. It's about the same for < 10 elements, about 10% faster with 100 elements and about 20% faster with 1000.